### ISSUE TITLE
System.OperationCanceledException when running orchestrators

### ISSUE THREAD

**Issue by anel-al** [Azure/azure-functions-durable-extension#1614]
Hi, 

Occasionally we are getting following exception when running our code for different orchestrations. I was not able to pinpoint cause by looking into our code. Could you please provide more information whether those are platform exceptions or something caused by user code?

With regards,

```
System.OperationCanceledException:
   at System.Threading.CancellationToken.ThrowOperationCanceledException (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.ServiceBus.Bindings.ServiceBusBinding+<BindAsync>d__10.MoveNext (Microsoft.Azure.WebJobs.ServiceBus, Version=3.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Triggers.TriggeredFunctionBinding`1+<BindCoreAsync>d__8.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Triggers\TriggeredFunctionBinding.csMicrosoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: 97)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Triggers.TriggeredFunctionBinding`1+<BindAsync>d__6.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Triggers\TriggeredFunctionBinding.csMicrosoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: 33)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+ParameterHelper+<BindAsync>d__31.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.csMicrosoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: 863)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<ExecuteWithLoggingAsync>d__19.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.csMicrosoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: 234)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<TryExecuteAsyncCore>d__16.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35Microsoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.csMicrosoft.Azure.WebJobs.Host, Version=3.0.23.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: 97)
```

**Comment by cgillum** [comment]
Hi @anel-al, it looks like this exception is being caused by the Service Bus binding in one of your functions. I could tell from this line in the callstack you provided:

> at Microsoft.Azure.WebJobs.ServiceBus.Bindings.ServiceBusBinding+<BindAsync>d__10.MoveNext (Microsoft.Azure.WebJobs.ServiceBus, Version=3.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)

This does not appear to be related to Durable Functions.

Since this is an `OperationCanceledException`, then it could mean that the binding operation you are performing is getting canceled, possibly due to a host shutdown. You can determine whether the host is shutting down by investigating the logs generated by the Functions host in Application Insights. If your app is a .NET-based app, you can even subscribe to host shutdown events by binding to a `CancellationToken` as described here: https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-class-library?tabs=v2%2Ccmd#cancellation-tokens

**Comment by anel-al** [comment]
Thank you @cgillum 

Based on your reply, I have additional question. There are two locations in activity where ServiceBus related code exists. Please see snippet bellow:

`
public async Task Run(
            [ActivityTrigger] PostEmailArguments arguments,
            ILogger logger,
            [ServiceBus("%SetInternalStatusRequestOutTopic1%", Connection = "ServiceBusConnection", EntityType = EntityType.Topic)] IAsyncCollector<InternalEvent> serviceBusTopic1,
            [ServiceBus("%SetInternalStatusRequestOutTopic2%", Connection = "ServiceBusConnection", EntityType = EntityType.Topic)] IAsyncCollector<InternalEvent> serviceBusTopic2,
            [ServiceBus("%SetInternalStatusRequestOutTopic3%", Connection = "ServiceBusConnection", EntityType = EntityType.Topic)] IAsyncCollector<InternalEvent> serviceBusTopic3
            )
        {
`

and location in activity body where message is pushed to the topic. For example:

`await serviceBusTopic1.AddAsync(internalEvent);`

Trying to figure out whether the cause for the exception is declaration or actual call.

Just to make sure, when orchestration calls this activity (with retry) and it fails during binding (construction) phase before it executes any user code, what would be default behavior? Is this activity call retried as configured and fails on last failure or it throws exception and fails orchestration immediately? 

With regards,
Anel

**Comment by cgillum** [comment]
@anel-al based on my understanding of how the `BindAsync` method works for binding extensions, I think this is happening right before your function code starts executing. This line in particular looks like it might be the source of the exception, based on some quick searching I did:

https://github.com/Azure/azure-functions-servicebus-extension/blob/f17eef19a992f927a5fd09aef62a47515dfb7cd5/src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Bindings/StringToServiceBusEntityConverter.cs#L40

In other words, I think this is happening in the function "declaration" as you described.

> when orchestration calls this activity (with retry) and it fails during binding (construction) phase before it executes any user code, what would be default behavior?

If an exception is thrown in the user code, we definitely handle it and do retries accordingly. However, when it's outside of user code we have to handle it differently. I believe the code for handling these cases is the following:

https://github.com/Azure/azure-functions-durable-extension/blob/faa0ccd9e5a6596365e8109c2ce6c4745e19c67d/src/WebJobs.Extensions.DurableTask/Listener/TaskActivityShim.cs#L86-L96

As described in the comments, we treat this as an internal failure and attempt to do an automatic retry without sending any result back to the orchestration. This means we _should_ be retrying the execution regardless of whether the orchestration configured a retry policy or not. The orchestration will not see this failure and will keep waiting for a response from the user code. In this case, you should see a "function aborted" trace message in Application Insights, which is reserved for cases like this (host shutdowns, binding failures, etc.) that are assumed to be transient.

**Comment by anel-al** [comment]
So if I understand you correctly, beside exception in logs, there would be no data loss due to failed (activity) execution? Activity would be executed again by another instance as a result of system retry or retry policy where applicable? 

What about orchestration with the same binding and this type of error? Is it also covered by platform?

Thank you @cgillum

**Comment by cgillum** [comment]
@anel-al yes, there would be no data loss, the activity would be retried automatically by another instance, and orchestrations have similar internal error handling policies in place (though you shouldn't be using bindings that do I/O in an orchestrator function).

**Comment by anel-al** [comment]
So, dead-lettering when orchestration fails message processing should be done via dedicated activity and not directly from orchestration?

Thanks for all the insights @cgillum

**Comment by cgillum** [comment]
Are you asking about dead-lettering for internal Durable Functions queue messages? There is no dead-lettering for internal failures in Durable Functions. The extension will just keep retrying until the internal failure is resolved.

**Comment by anel-al** [comment]
I was referring to the user codebase specific dead lettering. For example to dead letter job processed by orchestration in case of failure. If I understood correctly, this should be done via additional activity and not by using binding directly in orchestration due to durable functions I/O limitation?

Thanks

**Comment by cgillum** [comment]
I'll go ahead and close this issue since I think the main concerns have been addressed, but let me know if there's anything else I can help clarify.

**Comment by k290** [comment]
Adding to this in case someone else comes across the same issue.

I spent weeks debugging a problem where we were getting the same OperationCanceled exception on all of our HTTP Trigger function. endpoints with a ServiceBus binding output.  Posting for the benefit of others who may experience the same issue.

This was a misleading exception. Although it appeared that the servicebus binding was failing it was actually a completely different function crashing the host.

In Application Insights I was able to find a host shutdown trace and a corresponding exception. The exception related to one of our other functions crashing the host due to a blob binding connection string having invalid characters (In particular it was not properly hydrated on our deployments so looked like $(BlobConnection).

This caused the host to continually crash and spin up resulting in the OperationCanceled exception on our other HTTP trigger functions being called. 

Incredibly frustrating and time-consuming to debug so hopefully others who end up here may see this.
